app/components/autocomplete/channel_mention/index.js

```diff
@@ -30,17 +52,32 @@ function mapStateToProps(state, ownProps) {
         }
     }
 
-    const autocompleteChannels = {
-        myChannels: getMyChannels(state).filter((c) => c.type !== General.DM_CHANNEL && c.type !== General.GM_CHANNEL),
-        otherChannels: getOtherChannels(state)
+    const match = postDraft.substring(0, cursorPosition).match(regex);
+    let matchTerm = null;
+    if (match) {
+        matchTerm = isSearch ? match[1] : match[2];
+    }
+
+    const opts = {
+        matchTerm,
+        isSearch
     };
 
+    if (isSearch) {
+        filterPublicChannels(state, {...opts, array: publicChannels});
```

Comment:
I'm not going to repeat my other comments from at_mention/index.js, but most of that applies here as well